base/special/trig.jl

```diff
@@ -175,6 +175,108 @@ function asin(x::T) where T<:Union{Float32, Float64}
     return asin_kernel(t, x)
 end
 
+# atan methods
+ATAN_1_O_2_HI(::Type{Float64}) = 4.63647609000806093515e-01 # atan(0.5).hi
+ATAN_2_O_2_HI(::Type{Float64}) = 7.85398163397448278999e-01 # atan(1.0).hi
+ATAN_3_O_2_HI(::Type{Float64}) = 9.82793723247329054082e-01 # atan(1.5).hi
+ATAN_INF_HI(::Type{Float64}) = 1.57079632679489655800e+00 # atan(Inf).hi
+
+ATAN_1_O_2_HI(::Type{Float32}) = 4.6364760399f-01 # atan(0.5).hi
+ATAN_2_O_2_HI(::Type{Float32}) = 7.8539812565f-01 # atan(1.0).hi
+ATAN_3_O_2_HI(::Type{Float32}) = 9.8279368877f-01 # atan(1.5).hi
+ATAN_INF_HI(::Type{Float32}) = 1.5707962513f+00 # atan(Inf).hi
+
+ATAN_1_O_2_LO(::Type{Float64}) = 2.26987774529616870924e-17 # atan(0.5).lo
+ATAN_2_O_2_LO(::Type{Float64}) = 3.06161699786838301793e-17 # atan(1.0).lo
+ATAN_3_O_2_LO(::Type{Float64}) = 1.39033110312309984516e-17 # atan(1.5).lo
+ATAN_INF_LO(::Type{Float64}) = 6.12323399573676603587e-17 # atan(Inf).lo
+
+ATAN_1_O_2_LO(::Type{Float32}) = 5.0121582440f-09  # atan(0.5).lo
+ATAN_2_O_2_LO(::Type{Float32}) = 3.7748947079f-08  # atan(1.0).lo
+ATAN_3_O_2_LO(::Type{Float32}) = 3.4473217170f-08  # atan(1.5).lo
+ATAN_INF_LO(::Type{Float32}) = 7.5497894159f-08  # atan(Inf).lo
+
+ATAN_LARGE_X(::Type{Float64}) = 2.0^66 # seems too large? 2.0^60 gives the same
+ATAN_SMALL_X(::Type{Float64}) = 2.0^-27
+ATAN_LARGE_X(::Type{Float32}) = 2.0f0^26
+ATAN_SMALL_X(::Type{Float32}) = 2.0f0^-12
+
+atan_p(z::Float64, w::Float64) = z*@horner(w,
+     3.33333333333329318027e-01,
+     1.42857142725034663711e-01,
+     9.09088713343650656196e-02,
+     6.66107313738753120669e-02,
+     4.97687799461593236017e-02,
+     1.62858201153657823623e-02)
+atan_q(w::Float64) = w*@horner(w,
+     -1.99999999998764832476e-01,
+     -1.11111104054623557880e-01,
+     -7.69187620504482999495e-02,
+     -5.83357013379057348645e-02,
+     -3.65315727442169155270e-02)
+atan_p(z::Float32, w::Float32) = z*@horner(w, 3.3333328366f-01,  1.4253635705f-01, 6.1687607318f-02)
+atan_q(w::Float32) = w*@horner(w, -1.9999158382f-01, -1.0648017377f-01)
+@inline function atan_pq(x)
+    x² = x*x
+    x⁴ = x²*x²
+    # break sum from i=0 to 10 aT[i]z**(i+1) into odd and even poly
+    atan_p(x², x⁴), atan_q(x⁴)
+end
+function atan(x::T) where T<:Union{Float32, Float64}
+    # Method
+    #   1. Reduce x to positive by atan(x) = -atan(-x).
+    #   2. According to the integer k=4t+0.25 chopped, t=x, the argument
+    #      is further reduced to one of the following intervals and the
+    #      arctangent of t is evaluated by the corresponding formula:
+    #
+    #      [0,7/16]      atan(x) = t-t^3*(a1+t^2*(a2+...(a10+t^2*a11)...)
+    #      [7/16,11/16]  atan(x) = atan(1/2) + atan( (t-0.5)/(1+t/2) )
+    #      [11/16.19/16] atan(x) = atan( 1 ) + atan( (t-1)/(1+t) )
+    #      [19/16,39/16] atan(x) = atan(3/2) + atan( (t-1.5)/(1+1.5t) )
+    #      [39/16,INF]   atan(x) = atan(INF) + atan( -1/t )
+    #
+    #  If isnan(x) is true, then the nan value will eventually be passed to
```

Comment:
I don't have the  time to go through the code, but as mentioned in slack I don't think this is a good idea, exceptions are best checked at the start of the method, even if by chance they propagate. There are different kinds of NaN (not that Julia respects those in general), but if the internal functions change, NaNs may no longer propagate and the check is trivially computationally.

Reply:
I get what you're saying, but under the assumption that no one deletes the tests, it will most certainly get caught if this changes behavior.

Reply:
I still don't think this is the best thing to do, going beyond the the fact that the test would catch this.   at a more basic level, exceptional input should be throw at the start.

Reply:
I'm still not sure I understand your objection: can you elaborate here?

Reply:
- I.e. add back the `isnan` check at the start of the method.

If the user inputs a NaN, the method should return/throw this immediately, since exceptions should be checked at the start., instead of allowing the NaN to propagate and be returned.

Reply:
Also helps debugging and custom exception handling

Reply:
Sorry, didn't see these comments when posting https://github.com/JuliaLang/julia/pull/23383#issuecomment-339095859 

I've said it earlier: I don't have strong feelings here, so I'll just do as told :)

Reply:
Let's go ahead and add it.

Reply:
> If the user inputs a NaN, the method should return/throw this immediately, since exceptions should be checked at the start., instead of allowing the NaN to propagate and be returned.
  
Why? There is a test checking this.

Reply:
It's very unusual to put  exception checks at the end of the function. It's conventional to throw immediately, otherwise   you are doing useless computations just to allow the NaN to pass through.  This is the norm for all  other math code  

1. Check exceptions 
2. Compute on sanitized input 

Vs.
1 Compute on sanitized input 
2. Check exceptions   (wasted effort)

Thus we should follow the original openlibm code and  leave the check at the start.

Reply:
> otherwise you are doing useless computations just to allow the NaN to pass through

Worrying that your NaNs are not taking the fastest path doesn't really seem worthwhile.

Reply:
With that logic all the exceptional checks should be moved to the bottom, but this is never done in any math code. They are typically filtered first. It shouldn't take 10-20 ns to throw a `NaN` back or an `Inf` or similar.

Reply:
In addition, the cost of the `isnan` check is miniscule in comparison to the cost of the function evaluation.